From f6359dfcbec5be9ad136af593dde42ab09d241f0 Mon Sep 17 00:00:00 2001
From: Moria Zuberi <moria.zuberi@tandemg.com>
Date: Mon, 18 Oct 2021 14:53:04 +0300
Subject: [PATCH] [SW-55777] gaudi2: gcm: fix: overload for onix

1. Update code according to LKD changes.
2. Overload clock-gating for ONIX with empty
   implementation since ONIX does not have any
   clock-gating peripherals and trying to write
   to these registers causes preboot to hang.

Change-Id: I261c7a7efe8f7f38c3064719ad42a77ab90656bb
Signed-off-by: Moria Zuberi <moria.zuberi@tandemg.com>
---

diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 32241ce..6df4148 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -374,15 +374,12 @@
 
 static void gaudi2_enable_all_tpcs_clk_gating(void)
 {
-	int dcore, inst, instances, tpc_seq = -1;
+	int dcore, inst, tpc_seq;
 	u32 offset;
 
 	for (dcore = 0; dcore < NUM_OF_DCORES; dcore++) {
-		instances = (dcore == 0) ?
-				NUM_DCORE0_TPC : NUM_OF_TPC_PER_DCORE;
-
-		for (inst = 0; inst < instances; inst++) {
-			tpc_seq++;
+		for (inst = 0; inst < NUM_OF_TPC_PER_DCORE; inst++) {
+			tpc_seq = dcore * NUM_OF_TPC_PER_DCORE + inst;
 
 			if (!(TPC_ENABLED_MASK & BIT(tpc_seq)))
 				continue;
@@ -392,6 +389,11 @@
 			gaudi2_enable_tpc_clk_gating(dcore, inst, offset);
 		}
 	}
+
+	/* special check for PCI TPC (DCORE0_TPC6) */
+	offset = DCORE_TPC_OFFSET * (NUM_DCORE0_TPC - 1);
+	if (TPC_ENABLED_MASK & BIT(TPC_ID_DCORE0_TPC6))
+		gaudi2_enable_tpc_clk_gating(0, NUM_DCORE0_TPC - 1, offset);
 }
 
 void gaudi2_init_mme_clk_gating(void)
@@ -399,7 +401,7 @@
 	int i, j, dcore_offs = 0;
 
 	for (i = 0; i < NUM_OF_DCORES; i++) {
-		if (!(MME_MASK & (1 << i)))
+		if (!(MME_MASK & BIT(i)))
 			continue;
 
 		dcore_offs = i * DCORE_OFFSET;
@@ -487,7 +489,7 @@
 
 	for (i = 0; i < NUM_OF_DCORES; i++) {
 		for (j = 0; j < NUM_OF_VDEC_PER_DCORE; j++) {
-			if (!(DECODER_ENABLED_MASK & (1 << ((i * 2) + j))))
+			if (!(DECODER_ENABLED_MASK & BIT((i * 2) + j)))
 				continue;
 
 			reg = mmDCORE0_VDEC0_BRDG_CTRL_CGM_DISABLE +
@@ -497,7 +499,7 @@
 	}
 }
 
-void hl_clock_gating_init(void)
+__weak void hl_clock_gating_init(void)
 {
 	gaudi2_enable_all_tpcs_clk_gating();
 
diff --git a/arch/gaudi2/common/onix.c b/arch/gaudi2/common/onix.c
index 61cb7cf..3f412d1 100644
--- a/arch/gaudi2/common/onix.c
+++ b/arch/gaudi2/common/onix.c
@@ -96,3 +96,5 @@
 void hl_ts_shutdown(void) {}
 
 void preboot_update_bar_sizes(void) {}
+
+void hl_clock_gating_init(void) {}
